### PR TITLE
DM-14534: Fix measurementInvestigationLib.makeRerunCatalog parent keys

### PR DESCRIPTION
**Pull request by taranu**
Adds two options to makeRerunCatalog to either resetParents (default=True) to zero or addParents (default=False), so that child objects with parents outside of the rerunCatalog aren't skipped by runPlugins.

### CODE REVIEW

File: doc/ReconstructingMeasurements.rst [hunk 1]
```diff
@@ -84,3 +84,58 @@ id for each object.
...
+
+
+At the moment, measTask.runPlugins will not run on any child
+objects if their parents are not also in the catalog. This issue can be
+solved in several ways, two of which are options in makeRerunCatalog. The
+``resetParents`` flag (True by default) will reset the parent keys of any
+children whose parents are not included in the idList, effectively turning
+them into parents. The ``addParents`` flag (False by default) will add these
+parents to the idList. Some timed examples follow:
+
```
Comment (natelust) [inline doc/ReconstructingMeasurements.rst]: In the example below, all the timing code is unrelated to the topic of this document, and obfuscates the point being made. Examples should have the bare minimum amount of extra utility and concepts to demonstrate a point. 

Consider instead counting the number of nans (or zeros if that is the default) in a column (pick which ever makes sense)  before and after turning on resetParents, to show that the number decreases. You can also show that before turning on reset parents the number of bad values is equal to the number of sources that are children. To demonstrate adding parents, you can show the number of bad values as above, but you can also show the length of the inputId list vs the length of the output rerun catalog and how the later is now larger.

File: python/lsst/meas/base/measurementInvestigationLib.py [hunk 2]
```diff
@@ -97,13 +98,30 @@ def makeRerunCatalog(schema, oldCatalog, idList, fields=None):
...
     """
 
+    if not isinstance(schema, Schema):
+        raise RuntimeError("schema must be an instance of "
+                           "lsst.afw.table.Schema")
+
+    if not isinstance(oldCatalog, SourceCatalog):
+        raise RuntimeError("oldCatalog must be an instance of "
+                           "lsst.afw.table.SourceCatalogiterable")
+
```
[the remark this answers is not in the document]
Comment (taranu) [reply python/lsst/meas/base/measurementInvestigationLib.py]: Setting resetParents=True will have no effect if addParents=True, so it seems excessive to raise an exception given clearer documentation for resetParents.

File: python/lsst/meas/base/measurementInvestigationLib.py [hunk 3]
```diff
@@ -113,13 +131,30 @@ def makeRerunCatalog(schema, oldCatalog, idList, fields=None):
         if entry not in schema:
             schema.addField(oldCatalog.schema.find(entry).field)
 
+    # It's likely better to convert to a list and append
+    idList = list(idList)
```
[the remark this answers is not in the document]
Comment (taranu) [reply python/lsst/meas/base/measurementInvestigationLib.py]: :+1:

File: python/lsst/meas/base/measurementInvestigationLib.py [hunk 4]
```diff
@@ -113,13 +131,30 @@ def makeRerunCatalog(schema, oldCatalog, idList, fields=None):
...
+
+    if addParents:
+        lenIdList = len(idList)
```
[the remark this answers is not in the document]
Comment (taranu) [reply python/lsst/meas/base/measurementInvestigationLib.py]: :+1:

File: python/lsst/meas/base/measurementInvestigationLib.py [hunk 5]
```diff
@@ -113,13 +131,30 @@ def makeRerunCatalog(schema, oldCatalog, idList, fields=None):
...
+
     measCat = SourceCatalog(schema)
     for srcId in idList:
         oldSrc = oldCatalog.find(srcId)
         src = measCat.addNew()
         src.setId(srcId)
         src.setFootprint(oldSrc.getFootprint())
-        src.setParent(oldSrc.getParent())
+        parent = oldSrc.getParent()
+        if parent != 0 and resetParents and parent not in idList:
```
[the remark this answers is not in the document]
Comment (taranu) [reply python/lsst/meas/base/measurementInvestigationLib.py]: :+1:

File: doc/ReconstructingMeasurements.rst (hunk 1 above)
Comment (taranu) [reply doc/ReconstructingMeasurements.rst]: :+1:

File: doc/ReconstructingMeasurements.rst [hunk 6]
```diff
@@ -83,59 +83,42 @@ id for each object.
...
 
-    # Running the parents as well takes a little longer
+    # Running the parents as well with addParents=True takes a little longer
     newSrcCatalog = makeRerunCatalog(
-        schema, srcCat, idsToRerun, fields=fields, addParents=True,
-        resetParents=False
+        schema, srcCat, idsToRerun, fields=fields, addParents=True
     )
-    print("New (addParents=True, resetParents=False): ", timeRunPlugins())
+    measTask.runPlugins(rebuildNoiseReplacer(exposure, srcCat), newSrcCatalog, exposure)
```
Comment (natelust) [inline doc/ReconstructingMeasurements.rst]: My only last comment will be to add a code comment explaining the expected output, aka with parents more sources will be in the output, like you did in the above two cases

**Pull request merged**